source/solvers/postprocessing_cfd.cc

```diff
@@ -423,6 +423,105 @@ calculate_kinetic_energy<3, TrilinosWrappers::MPI::BlockVector>(
   const Mapping<3> &                        mapping);
 
 
+template <int dim, typename VectorType>
+double
+calculate_apparent_viscosity(
+  const DoFHandler<dim> &               dof_handler,
+  const VectorType &                    evaluation_point,
+  const Quadrature<dim> &               quadrature_formula,
+  const Mapping<dim> &                  mapping,
+  const Parameters::PhysicalProperties &physical_properties)
+{
+  double integral_viscosity_x_shear_rate = 0;
+  double integral_shear_rate             = 0;
+  double shear_rate_magnitude;
+  double viscosity;
+  // Cast rheological model to either a Newtonian model or one of the
+  // non Newtonian models according to the physical properties
+  std::shared_ptr<RheologicalModel<dim>> rheological_model =
+    RheologicalModel<dim>::model_cast(physical_properties);
+
+  const FESystem<dim, dim> fe = dof_handler.get_fe();
+  FEValues<dim>            fe_values(mapping,
+                          fe,
+                          quadrature_formula,
+                          update_values | update_gradients |
+                            update_quadrature_points | update_JxW_values);
+
+  const FEValuesExtractors::Vector velocities(0);
+
+  const unsigned int n_q_points = quadrature_formula.size();
+
+  std::vector<Tensor<2, dim>> present_velocity_gradients(n_q_points);
+  double                      domain_volume =
+    GridTools::volume(dof_handler.get_triangulation(), mapping);
```

Comment:
Instead of using this function, I would simply integrate the JwX in the following loop as well this will eliminate a loop over the cells.

Reply:
I'm not sure I understand this comment, let me fix something with your previous comment and recomment that.